sdk/client/src/move_deserialize.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, Result};
+use diem_json_rpc_types::views::{AccountStateWithProofView, EventWithProofView};
+use diem_types::{
+    account_state::AccountState,
+    account_state_blob::AccountStateBlob,
+    contract_event::{ContractEvent, EventWithProof},
+};
+use move_core_types::move_resource::MoveResource;
+use serde::de::DeserializeOwned;
+use std::convert::TryFrom;
+
+/// Wrapper for a deserialized Move event and its containing `ContractEvent`
+#[derive(Debug, Clone)]
+pub struct Event<T: MoveResource + DeserializeOwned> {
+    /// The deserialized event type
+    data: T,
+    event: ContractEvent,
+}
+
+/// Deserialize and return the Move events of type `T` in `events`
+pub fn get_events<T: MoveResource + DeserializeOwned>(
+    events: Vec<EventWithProofView>,
+) -> Result<Vec<Event<T>>> {
+    let events_with_proof: Vec<EventWithProof> = events
+        .into_iter()
+        .map(|e| {
+            bcs::from_bytes::<EventWithProof>(e.event_with_proof.inner()).map_err(Error::decode)
+        })
+        .collect::<Result<Vec<EventWithProof>>>()?;
+    events_with_proof
+        .into_iter()
+        .map(|e| {
+            bcs::from_bytes::<T>(e.event.event_data())
+                .map(|data| Event {
+                    data,
+                    event: e.event,
+                })
+                .map_err(Error::decode)
+        })
+        .collect::<Result<Vec<Event<T>>>>()
+}
+
+fn get_account_state(
+    account_state_with_proof: AccountStateWithProofView,
+) -> Result<Option<AccountState>> {
+    let account_opt = account_state_with_proof.blob;
+    if let Some(account) = account_opt {
+        let account_state_blob: AccountStateBlob =
+            bcs::from_bytes(account.inner()).map_err(Error::decode)?;
+        return Ok(Some(
+            AccountState::try_from(&account_state_blob).map_err(Error::decode)?,
+        ));
+    }
+    Ok(None)
+}
+
+/// Deserialize and return the Move value of type `T` in `account_state_with_proof`
+/// Returns None if no resource of type `T` exists under `address`
+pub fn get_resource<T: MoveResource + DeserializeOwned>(
```

Comment:
Get resource looks like it requires `MoveResource` so these look like sane bounds to me.